GVFS/GVFS.Common/Database/PlaceholdersTable.cs

```diff
@@ -0,0 +1,202 @@
+            }
+        }
+
+        public void AddPlaceholderData(IPlaceholderData data)
+        {
+            if (data.IsFolder)
+            {
+                if (data.IsExpandedFolder)
+                {
+                    this.AddExpandedFolder(data.Path);
+                }
+                else if (data.IsPossibleTombstoneFolder)
+                {
+                    this.AddPossibleTombstoneFolder(data.Path);
+                }
+                else
+                {
+                    this.AddPartialFolder(data.Path);
+                }
+            }
+            else
+            {
+                this.AddFile(data.Path, data.Sha);
+            }
+        }
+
+        public void AddFile(string path, string sha)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.File, Sha = sha });
+            }
+        }
+
+        public void AddPartialFolder(string path)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.PartialFolder });
+            }
+        }
+
+        public void AddExpandedFolder(string path)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.ExpandedFolder });
+            }
+        }
+
+        public void AddPossibleTombstoneFolder(string path)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.PossibleTombstoneFolder });
+            }
+        }
+
+        public void Remove(string path)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Delete(command, path);
+            }
+        }
+
+        private static void Insert(IDbCommand command, PlaceholderData placeholder)
+        {
+            command.CommandText = "INSERT OR REPLACE INTO Placeholders (path, pathType, sha) VALUES (@path, @pathType, @sha);";
+            command.AddParameter("@path", DbType.String, placeholder.Path);
+            command.AddParameter("@pathType", DbType.Int32, (int)placeholder.PathType);

```

Comment:
`PlaceholderData.PathType` is an enumeration that is backed by a `byte`, but here we're casting it to `int`. Generally advice for enumerations is _do not specify a backing type_ unless there's a good reason.

Reply:
I will have to measure the memory and see.  We load all the placeholders into memory when we update them and if the user has most of the placeholders created ~3 million times 3 bytes is ~9MB that we save.

Reply:
`PlaceholderData` is a reference type and is being heap allocated. The `PlaceholderType PathType` member is a property, which will auto-create a private backing field.

> [..] is ~9MB that we save.

I'm not sure I know enough about how the CLR allocates to say if this would be true. Are instance fields packed or aligned to some n-byte boundary? In which order (does it matter)?

```csharp
class PlaceholderData
{
    private byte pathType;
    private int howAmIAligned; // 1 byte in or 4? 8?

    public byte PathType { get => pathType; set => pathType = value; }
}
```